File: Problems/Algorithms/994. Rotting Oranges/RottingOranges.kt
class Solution {
    fun orangesRotting(grid: Array<IntArray>): Int {
        val fresh = mutableSetOf<Pair<Int, Int>>()
        val rotten = ArrayDeque<Pair<Int, Int>>()
        
        for (i in 0..grid.size-1) {
            for (j in 0..grid[0].size-1) {
                if (grid[i][j] == 1) fresh.add(Pair(i, j))
                if (grid[i][j] == 2) rotten.add(Pair(i, j))
            }
        }
        
        var ans = 0
        while (!fresh.isEmpty() && !rotten.isEmpty()) {
            ans++
            
            val currSize = rotten.size
            for (k in 0..currSize-1) {
                val (i, j) = rotten.removeFirst()
                
                for (neighbor in arrayOf(Pair(i-1, j), Pair(i+1, j), Pair(i, j-1), Pair(i, j+1))) {
                    if (fresh.contains(neighbor)) {
                        fresh.remove(neighbor)
                        rotten.add(neighbor)
                    }
                }
            }
        }
        
        if (!fresh.isEmpty()) return -1
        return ans
    }
}

File: Problems/Algorithms/994. Rotting Oranges/rotting_oranges.py
class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        fresh = set()
        rotten = []

        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] == 1:
                    fresh.add((i, j))
                if grid[i][j] == 2:
                    rotten.append((i, j))

        ans = 0
        while len(fresh) > 0 and len(rotten) > 0:
            ans += 1

            curr_size = len(rotten)
            for k in range(curr_size):
                i, j = rotten.pop(0)

                for neighbor in [(i-1, j), (i+1, j), (i, j-1), (i, j+1)]:
                    if neighbor in fresh:
                        fresh.remove(neighbor)
                        rotten.append(neighbor)

        if len(fresh) > 0:
            return -1
        return ans

